simple/msg_data_integrity.c

```diff
@@ -0,0 +1,717 @@
+		goto err0;
+	}
+
+	ret = fi_fabric(fi->fabric_attr, &fab, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_fabric", ret);
+		goto err1;
+	}
+
+	ret = fi_domain(fab, fi, &dom, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_domain", ret);
+		goto err2;
+	}
+
+	ret = fi_endpoint(dom, fi, &ep, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_endpoint", ret);
+		goto err3;
+	}
+
+	ret = alloc_ep_res(fi);
+	if (ret)
+		goto err3;
+
+	ret = bind_ep_res();
+	if (ret)
+		goto err5;
+
+	ret = fi_connect(ep, fi->dest_addr, NULL, 0);
+	if (ret) {
+		FT_PRINTERR("fi_connect", ret);
+		goto err5;
+	}
+
+	rd = fi_eq_sread(cmeq, &event, &entry, sizeof entry, -1, 0);
+	if (rd != sizeof entry) {
+		if (rd == -FI_EAVAIL) {
+			rd = fi_eq_readerr(cmeq, &err, 0);
+			if (rd != sizeof(err)) {
+				FT_PRINTERR("fi_eq_sread", rd);
+			} else {
+				fprintf(stderr, "EQ report error %d %s\n", err.err,
+						fi_strerror(err.err));
+			}
+		} else {
+			FT_PRINTERR("fi_eq_sread", rd);
+		}
+		return (int) rd;
+	}
+
+	if (event != FI_CONNECTED || entry.fid != &ep->fid) {
+		fprintf(stderr, "Unexpected CM event %d fid %p (ep %p)\n",
+			event, entry.fid, ep);
+		ret = -FI_EOTHER;
+		goto err5;
+	}
+
+	fi_freeinfo(fi);
+	return 0;
+
+err5:
+	free_ep_res();
+err3:
+	fi_close(&dom->fid);
+err2:
+	fi_close(&fab->fid);
+err1:
+	fi_freeinfo(fi);
+err0:
+	return ret;
+}
+
+static int run(void)
+{
+	int i, ret = 0;
+
+	if (!opts.dst_addr) {
+		ret = server_listen();
+		if (ret)
+			return ret;
+	}
+
+	ret = opts.dst_addr ? client_connect() : server_connect();
+	if (ret) {
+		return ret;
+	}
+
+	if ((opts.user_options & FT_OPT_ITER) || (opts.user_options & FT_OPT_SIZE)) {
+		init_test(&opts, test_name, sizeof(test_name));
+		ret = run_test();
+		if (ret)
+			goto cleanup;
+	} else {
+		for (i = 0; i < TEST_CNT; i++) {
+			if (test_size[i].option > opts.size_option)
+				continue;
+			opts.transfer_size = test_size[i].size;
+			init_test(&opts, test_name, sizeof(test_name));
+			ret = run_test();
+			if (ret) {
+				goto cleanup;
+			}
+		}
+	}
+
+	ret = wait_for_completion(scq, max_credits - credits);
+	/* Finalize before closing ep */
+        ft_finalize(ep, scq, rcq, FI_ADDR_UNSPEC);
+cleanup:
+	fi_shutdown(ep, 0);
+	free_ep_res();
+	if (!opts.dst_addr)
+		free_lres();
+	fi_close(&dom->fid);
+	fi_close(&fab->fid);
+	return ret;
+}
+
+int main(int argc, char **argv)
+{
+	int op, ret;
+	opts = INIT_OPTS;
+
+	hints = fi_allocinfo();
+        if (!hints)
+                return EXIT_FAILURE;
+
+	while ((op = getopt(argc, argv, "c:aeh" CS_OPTS INFO_OPTS)) != -1) {
+		switch (op) {
+		case 'c':
+			allowed_error_count = atoi(optarg);
+			break;
+		case 'a':
+			detailed_output = 1;
+			break;
+		case 'e':
+			detailed_error = 1;
+			break;
+		default:
+			ft_parseinfo(op, optarg, hints);
+			ft_parsecsopts(op, optarg, &opts);
+			break;
+		case '?':
+		case 'h':
+			ft_csusage(argv[0], "Ping pong client and server using message endpoints.");
+			FT_PRINT_OPTS_USAGE("-c <count>", "allowed error count");
```

Comment:
I'm not sure I see the value of the `allowed_error_count`.  EP_MSG and EP_RDM endpoints are supposed to be fully reliable, so it's a real problem as soon as there is a single instance of data corruption.